DM-15098: Add Registry.getRegion(DataId)

pschella opened the pull request.

timj commented on python/lsst/daf/butler/registries/sqlRegistry.py:
```diff
@@ -992,6 +994,42 @@ def find(self, collection, datasetType, dataId):
+
+        Raises
+        ------
+        `ValueError`
```
Exception names in a Raises block are not backtick quoted.

timj commented on tests/test_sqlRegistry.py:
```diff
@@ -376,6 +377,17 @@ def testBasicTransaction(self):
+        region = lsst.sphgeom.ConvexPolygon((lsst.sphgeom.UnitVector3d(1, 0, 0),
+                                             lsst.sphgeom.UnitVector3d(0, 1, 0),
+                                             lsst.sphgeom.UnitVector3d(0, 0, 1)))
+        registry.addDataUnitEntry("SkyMap", {"skymap": "DummySkyMap", "sha1": bytes()})
```
I've just noticed we are insisting on this being a `sha1`. Would it generically be better to call this `hash` or something?

pschella replied:
Yes, but out of scope for this ticket.

timj commented on tests/test_sqlRegistry.py:
```diff
@@ -376,6 +377,17 @@ def testBasicTransaction(self):
+                                             lsst.sphgeom.UnitVector3d(0, 0, 1)))
+        registry.addDataUnitEntry("SkyMap", {"skymap": "DummySkyMap", "sha1": bytes()})
+        registry.addDataUnitEntry("Tract", {"skymap": "DummySkyMap", "tract": 0, "region": region})
+        dataId = {"skymap": "DummySkyMap", "tract": 0}
```
Can you also show that `tract: 1` doesn't get retrieved?

timj commented on python/lsst/daf/butler/registries/sqlRegistry.py:
```diff
@@ -992,6 +994,42 @@ def find(self, collection, datasetType, dataId):
+        Parameters
+        ----------
+        dataId : `dict`
+            A `dict` of `DataUnit` name, value pairs that label the `DatasetRef`
```
Is it confusing to call out to the `DataUnit` class docs here even though these arguments are not of that class?

pschella replied:
It is, and that should be fixed in multiple places. I'll do it on this ticket.

timj commented on python/lsst/daf/butler/registries/sqlRegistry.py:
```diff
@@ -992,6 +994,42 @@ def find(self, collection, datasetType, dataId):
+            dataUnit = self._schema.dataUnits.getByLinkName(key)
+            if dataUnit.regionColumn is not None:
+                if haveSpatialDataUnit:
+                    raise NotImplementedError("Cannot have more than one spatial region")
```
Please add a test for this condition.

timj reviewed (approved):
Minor comments regarding docstrings and test coverage but otherwise fine.

andy-slac commented on python/lsst/daf/butler/core/dataUnit.py:
```diff
@@ -514,3 +517,18 @@ def getPrimaryKeyNames(self, dataUnitNames):
+        -------
+        dataUnit : `DataUnit`
+            The corresponding `DataUnit` instance.
+        """
```
Add `Raises` section?

Commit:
Consistent docstring for DataId.

Commit:
Mark DataUnit(Region)s explicitly as spatial.

This makes it easier to make the distinction between
dataunits that have a table with a region column (most spatial
dataunits) and ones that don't but are still spatial (SkyPix).

Commit:
Implement Registry.getRegion.
